src/Analyser/MutatingScope.php

```diff
@@ -3466,23 +3470,29 @@ public function specifyExpressionType(Expr $expr, Type $type, Type $nativeType):
 					if ($dimType instanceof ConstantIntegerType) {
 						$types[] = new StringType();
 					}
+
 					$scope = $scope->specifyExpressionType(
 						$expr->var,
 						TypeCombinator::intersect(
 							TypeCombinator::intersect($exprVarType, TypeCombinator::union(...$types)),
 							new HasOffsetValueType($dimType, $type),
 						),
 						$scope->getNativeType($expr->var),
+						$scope->hasExpressionType($expr->var), // keep certainty
 					);
 				}
 			}
 		}
 
+		if ($certainty === null) {
+			$certainty = TrinaryLogic::createYes();
```

Comment:
Please reject `->no()` values, they don't make sense here. `assignVariable` also does it.